### PR TITLE
Project introduction: initial app module

### LINKED ISSUE #21: Implement StoryProgressController
Users will make progress on individual stories when they play them. There should be a system that can be used to record and retrieve this state based on the story being played (by ID) and profile ID of the user playing it. Note that this system is storing based on individual exploration being played in the context of a story.

The actual storage necessary here is just completed exploration IDs that can be cross-referenced with the exploration IDs contained in a story.

### PR DESCRIPTION
**Pull request by BenHenning**
Initial app module introduction.
    
This is based on the empty project generated by Android Studio. It includes an adaptive Oppia app logo based on the prototype app. It also includes a shared test that can run both with Robolectric and Espresso. The initial screen is a basic view with an action bar and a welcome string that's being verified in two different ways in tests.

This check-in also includes the necessary Gradle and Android Studio/IntelliJ files needed to load and build the project, and .gitignore filters to avoid pulling in the unneeded pieces.

### CODE REVIEW

File: .idea/codeStyles/Project.xml
Comment (seanlip) [inline .idea/codeStyles/Project.xml]: Should we be adding newlines at EOFs, here and in other files?

File: app/.gitignore
Comment (seanlip) [inline app/.gitignore]: Why do we need one .gitignore for app/ and one .gitignore for the root?

**Review by seanlip — APPROVED** [review]
LGTM, just a couple small comments. Thanks!

I'll try and look into why I can't be assigned -- will get back on this.

**Comment by seanlip** [comment]
Ah I think it was a permissions issue. It should probably work now?

I added @rt4914 too.

**Comment by rt4914** [comment]
Got access. Thanks

File: .idea/codeStyles/Project.xml
Comment (BenHenning) [reply .idea/codeStyles/Project.xml]: Done, though IntelliJ doesn't seem to have a way to enforce this so we might not be able to easily keep this consistent without a GitHub check.

File: app/.gitignore
Comment (BenHenning) [reply app/.gitignore]: We don't. This is generated by IntelliJ, but I agree it's unnecessary. Cleaned up.

**Comment by BenHenning** [comment]
Thanks @seanlip!

**Comment by rt4914** [comment]
Hello Chantel,

Can you please share custom-icons which you created for tabs in topic.

Reference Link:
https://xd.adobe.com/spec/e2239cf4-9cde-4c08-5296-25316c1f0a14-9412/screen/19cfbacf-854c-4c7d-8691-3b3d117e1866/Home-Page-1-Start-Learning-Pre-Scroll-3

Best,
Rajat

**Comment by rt4914** [comment]
Hi Rajat,

No problem! Here are the SVG files of the Play, Train, and Review icons. I
believe the Overview icon can be found on the Material Icons list
<https://material.io/resources/icons/?search=info&style=outline> under
"outline" and filtered by the term "info." If there are any other custom
icons that I find in my designs, I will let you know and send them to you
as well. Let me know if you need anything else!

Best,
Chantel

**Comment by rt4914** [comment]
Hi Rajat,

I also found some custom icons that I made for the Navigation Drawer (aka
side menu). These icons should follow the "outline" theme, and I will
update the prototype to reflect this. The material icons can be found by
searching for the following:

   - Home: home
   - Preferences: tune
   - Help: *attached to email*
   - Downloads: folder_open
   - Profile Chooser: *attached to email*
   - Administrator Settings: settings_application
      - note -- In my prototype, I used only a cog as the Administrator
      Settings icon. I am not sure why, but the icon is only available as a cog
      within a box. Just for consistency, we will use the icon with the box.

Best,
Chantel

**Comment by rt4914** [comment]
Awesome, thanks a lot.

**Comment by rt4914** [comment]
Hello everyone,

Follow these below mentioned PR guidelines if possible:
1. Each PR should have roughly around 4-5 files to review.
2. Each PR should have commit names, branch name, PR title, as per
guidelines mentioned on Oppia-Web Wik
<https://github.com/oppia/oppia-android/wiki>i.
3. Once your PR is ready mention Reviewers and Assignees. Also, The
Assignees list will keep on changing depending who needs to review that PR.
4. As a reviewer make sure that once you review the PR, you assign it back
to the code-owner.

Tip: If your PR takes long and a lot of conversations are taking place on
that PR, you should either split your PR in multiple parts, so that each PR
has less files to review or you should start reviewing PRs of other
developers that will help you improve your PR.

Best,
Rajat

**Comment by rt4914** [comment]
Hi Ben,

Bangalore Team has internally decided on a new PR guideline as per our
discussion in last 1:1, i.e., PRs should generally have 3-4 files for
review.

Assuming the case of TopicTrainFragment low-fi (Sample PR: #197), which
contains a list of skills and on user selection, Submit button will get
activated.

Now this PR was earlier on #197, but now I have recreated this same PR on
#200, #202, #203, #204.

1. *#200 - topic-train-low-fi-part-1 (7 files to review)* -> This PR
contains all the xml files except layout files, like colors, strings,
drawables, styles, etc. We expect generally such PRs will be merged
directly once it gets reviewed by one of the team-members.
2. *#202 - topic-train-low-fi-part-2 (3 files to review)* -> This PR
contains the implementation of kotlin and xml code for recyclerview item
usage like SkillSelectionAdapter.kt, topic_train_skill_view.xml (Recycler
View Item Layout) and SkillInterface for transferring user skill selection
information.
3. *#203 - topic-train-low-fi-part-3 (4 files to review) *-> This PR
contains final implementation of this issue and combines everything in
previous two PRs.
4. *#204 - topic-train-low-fi-part-4 (1 file to review)* -> This final PR
contains test-case for this issue.

Also, #200 has already been merged with approval of Veena and Nikita.

Process behind such PRs.
1. The base PR should mostly be related to xml files which can be reviewed
and merged with approval from Bangalore team members without taking it
ahead to Ben.
2. Second PR can be mostly discrete kotlin and xml files which will be used
in the issue but can be merged independently like item-layout, Interface,
Adapter, etc.
3. Next few PRs should be towards the final implementation and combining
everything for full functionality.
4. Separate PR for test-cases, this has been decided keeping two-points in
mind

a. Even if there are any changes in this PR, the developer won't get stuck
and can work on other issues in parallel, assuming that the other PRs have
already been merged.

b. Also, I think once the Bangalore team becomes more comfortable with
testing, this PR can be merged with reviews from other members from
Bangalore team.


Note for such PRs:
1. Delete all branches once all the PRs have been approved.
2. Your first-branch will base-off from develop, second branch will
base-off from first-branch and so on.
3. All branches should have part-number mentioned to them, like
topic-train-low-fi-*part-1*.
4. All PR description should have links to other PR parts.
5. The PR final merge to develop should strictly follow hierarchy similar
to branches, meaning, your first PR should get merged first, then second,
then third and so on.

Feel free to improvise this process as you go along and also do share tips
on improvising this further with everyone.

Best,
Rajat

**Comment by rt4914** [comment]
Thanks Rajat.

This seems sensible to start with. A couple of points though:

   1. We should be careful about introducing XML ahead of Kotlin: it's
   quite likely the XML will change, and it's difficult to review these
   changes outside of the corresponding Java file. It's also difficult to
   review the Kotlin code without understanding the layouts. I actually think
   we should keep things in well-defined, standalone containers which means
   the XML should be kept together with its Kotlin counterpart code, since I
   find myself usually offering comments in both layout and Kotlin code.
   2. I'm generally not in favor of separating implementations from tests
   since I often use tests to verify that the implementation is correct, and
   the implementation to ensure that the right tests are being covered. I'm
   fine with us employing this process for now only for UI structure work
   items, and those work items should not be marked as fixed until their
   corresponding tests are completed.

It's completely fine for PRs to be larger than 4 files if all of the files
are comprise a single logical unit. Separating out the recycler view item
from the unit that contains the recyclerview is a good example of
conceptual separation, but splitting pieces within each of those makes it
harder to review that the unit as a whole is correct. I suggest we keep PRs
to a single logical unit, and only for UI structure work separate out the
tests (but still complete them as part of the work item).

Does this adjustment make sense?

Also, I'll review the existing PRs that are out as-is. :)

Thanks!
Ben

**Comment by rt4914** [comment]
Hi Ben,

Your adjustments do make sense. I will follow this improvised approach in
my next PRs.

Also, regarding XML files PR part-1 I meant that in part-1 xml files should
be out for review but in this layout-xml files won't be there. So
basically, this PR should deal with strings, colors, dimens, drawables only.

@Sean Lip <sean@seanlip.org> For now we are keeping this here, once we meet
our deadlines I guess we can then work integrating such things in
Oppia-Android Wiki.

Best,
Rajat

**Comment by rt4914** [comment]
Sounds good--thanks Rajat. And I agree that we should codify some of these
guidelines onto the wiki once we have a better idea of what the long-term
PR guidelines should be.

**Pull request merged**
